cudamapper/src/index_cache.cu

```diff
@@ -41,84 +41,164 @@ IndexCacheHost::IndexCacheHost(const bool reuse_data,
-                                                                        descriptor_of_index_to_cache.first_read(),
-                                                                        descriptor_of_index_to_cache.first_read() + descriptor_of_index_to_cache.number_of_reads(),
-                                                                        kmer_size_,
-                                                                        window_size_,
-                                                                        hash_representations_,
-                                                                        filtering_parameter_,
-                                                                        cuda_stream_);
-            // copy it to host memory
-            index_copy = claragenomics::cudamapper::IndexHostCopy::create_cache(*index,
-                                                                                descriptor_of_index_to_cache.first_read(),
-                                                                                kmer_size_,
-                                                                                window_size_,
-                                                                                cuda_stream_);
+            // check if this index is already cached in this cache
+            auto existing_cache = cache_to_edit.find(descriptor_of_index_to_cache);
+            if (existing_cache != cache_to_edit.end())
+            {
+                // index already cached
+                index_copy = existing_cache->second;
+                if (keep_on_device)
+                {
+                    index_on_device = index_copy->copy_index_to_device(allocator_, cuda_stream_);
+                }
+            }
+            else
+            {
+                // create index
+                index_on_device = claragenomics::cudamapper::Index::create_index(allocator_,
+                                                                                 *parser,
+                                                                                 descriptor_of_index_to_cache.first_read(),
+                                                                                 descriptor_of_index_to_cache.first_read() + descriptor_of_index_to_cache.number_of_reads(),
+                                                                                 kmer_size_,
+                                                                                 window_size_,
+                                                                                 hash_representations_,
+                                                                                 filtering_parameter_,
+                                                                                 cuda_stream_);
+                // copy it to host memory
+                index_copy = claragenomics::cudamapper::IndexHostCopy::create_cache(*index_on_device,
```

Comment:
Optional: I would drop the claragenomics::cudamapper prefix.

Reply:
done